README.md
# nsl-jist-swans
This program is based on [Jist/SWANS](http://jist.ece.cornell.edu/)  
BugFix and Add New function by [Network System Lab](http://www.nsl.ics.ritsumei.ac.jp/).

changelog.txt
===============================================================================
【タイトル】変更箇所
【作 成 日】07年06月22日
【改 訂 日】07年09月18日，07年09月17日, 07年11月12日，08年06月04日
　　　　　　08年09月10日，08年09月27日，08年11月07日，08年11月18日
===============================================================================

■はじめに
このファイルはJiST/SWANS ver.1.0.6の変更箇所を記述しています。


■変更箇所
----------------------------------------------------------------------

【パッケージ】jist.runtime.guilog
【ファイル名】GuiLog.java
【改  訂  日】2007/09/17
【変  更  点】934行目のコメントアウト
【説      明】デフォルトでエラー箇所となる部分。動作に支障なし

----------------------------------------------------------------------

【パッケージ】jist.runtime
【ファイル名】Constants.java
【改  訂  日】2007/09/17
【変  更  点】SENSITIVITY_DEFAULT，THRESHOLD_DEFAULTの値の変更
【説      明】送信半径、受信半径に関する値をの変更
　　　　　　　通信可能半径の導出式を以下に示す
　　　　　　　Pr ＝ Pt − Lp ＋ Gt ＋ Gr
　　　　　　　利得0dBのアイソトロピックアンテナで考えると、Lpは以下の式で求めることができる
　　　　　　　LB ＝ 10log10(4πd/λ)2 ＝ 20log10(4πd/λ) ＝ 20log10(4πdf/3×10^8)
　　　　　　　（Pr：受信電力 (dBm)、Pt：送信電力 (dBm)、Lp：伝搬損失 (dB)）
　　　　　　　（Gt：送信アンテナ利得 (dBi)、Gr：受信アンテナ利得 (dBi)） 

----------------------------------------------------------------------

【パッケージ】jist.swans.field
【ファイル名】Mobility.java
【改  訂  日】2007/09/17
【変  更  点】RandomWaypointの実装
【説      明】他の移動モデル同様に、引数を分解しinitメソッドに渡す

【改  訂  日】2008/09/10
【変  更  点】全てのモデルにおいて引数pausetimeの単位を統一，エラーメッセージの修正
【説      明】Teleportクラスのコンストラクタに引数として与えるpauseTimeの単位を秒に変更
　　　　　　　RandomWaypointのエラーメッセージの修正

----------------------------------------------------------------------

【パッケージ】jist.swans.field
【ファイル名】Field.java
【改  訂  日】2007/09/17
【変  更  点】startMobility(Integer, boolean)メソッドの追加
【説      明】booleanの値によって移動モデルに従うか、静止状態を保つかを決める
　　　　　　　一定の割合のノードを動かしたいときに利用

----------------------------------------------------------------------

*** 2007/10/08 802.11およびRadioNoiseについて検討する必要があります ***

【パッケージ】jist.swans.mac
【ファイル名】Mac802_11.java(検討の余地あり)
【改  訂  日】2007/09/17
【変  更  点】RTS/CTS処理の追加，シーケンスナンバー追加処理の変更
　　　　　　　受信時のシーケンスナンバー処理の変更
　　　　　　　プロミスキャスモードの受信時の処理変更
【説      明】RTS/CTSを利用できるように変更
　　　　　　　802.11の仕様に従い、再送時にシーケンスナンバーが加算されるのを修正
　　　　　　　受信時に上の層に渡す判定を厳格化
　　　　　　　プロミスキャスモードで受信時にACK応答時間待機してからネットワーク層に渡すよう修正

【改  訂  日】2007/09/19
【変  更  点】ブロードキャスト受信時処理修正
【説      明】DIFSが発生しないための応急処置

【改  訂  日】2007/10/01
【変  更  点】ブロードキャスト送信時処理修正
【説      明】送信後にリトライ値をリセット

【改  訂  日】2007/10/08
【変  更  点】ACK受信時処修正，CTS受信後処理修正，BO後の処理修正，送信前処理修正
【説      明】DIFSが発生しないための応急処置，CTS受信時にCWをリセット
　　　　　　　BO後にキャリアセンス処理を追加
　　　　　　　送信前に無線の状態を調べる処理を追加

【改  訂  日】2008/01/07
【変  更  点】データ受信処理修正
【説      明】データ受信時のシーケンスナンバー判定を修正

【改  訂  日】2008/06/04
【変  更  点】プロミスキャスモードの動作を修正
【説      明】データ受信時に受信完了までsleepする

【改  訂  日】2008/09/10
【変  更  点】timeoutの動作を一部変更 2007/10/08の処理の再修正
【説      明】case SBOにおいて、パケットの有無を確認してからキャリアセンス処理を行うように修正

【改  訂  日】2008/11/07
【変  更  点】timeoutの動作を一部変更 2007/10/08の処理および2008/09/10の処理を再修正
【説      明】送信直前のキャリアセンス処理において、ビジー時のretryを訂正

----------------------------------------------------------------------

【パッケージ】jist.swans.net
【ファイル名】NetIp.java
【改  訂  日】2007/09/17
【変  更  点】TTL減算処理変更，キューサイズ変更，フラッディング処理の追加、受信判定の変更
【説      明】TTLが0でも再送してしまう問題を修正
　　　　　　　キューのサイズを増加
　　　　　　　TCPおよびUDPメッセージ時のみフラッディング処理が可能
　　　　　　　フラッディングの際、既に送信済みでないかを判定

【改  訂  日】2007/11/12
【変  更  点】利用するMessageQueueのクラス変更
【説      明】インターフェース登録時に利用するMessageQueueのクラスを変更

【改  訂  日】2008/09/29
【変  更  点】一部変数名およびreceive()
【説      明】フラッディングの動作を変更
　　　　　　　2007/09/17のフラッディングの関する処理を削除
　　　　　　　初期状態ではフラッディングは無効になっています

【改  訂  日】2008/11/18
【変  更  点】receive()内の受信判定
【説      明】フラッディングの動作時の受信判定を修正
　　　　　　　2007/09/17の受信判定の変更を削除･変更

----------------------------------------------------------------------

【パッケージ】jist.swans.net
【ファイル名】BroadcastCache.java
【改  訂  日】2007/09/17
【変  更  点】新規設計
【説      明】フラッディングの際の送信済み判定処理用に新規作成

【改  訂  日】2008/09/29
【変  更  点】削除
【説      明】FloodingInaterfaceに変更

----------------------------------------------------------------------

【パッケージ】jist.swans.net
【ファイル名】FloodingInterface.java
【改  訂  日】2008/09/29
【変  更  点】新規設計
【説      明】フラッディング用に設計　インタフェースおよび実装したクラス

----------------------------------------------------------------------

【パッケージ】jist.swans.net
【ファイル名】MessageQueue.java
【改  訂  日】2007/09/17
【変  更  点】キューサイズを超過した場合の処理を変更
【説      明】キューサイズを超過した場合，登録しようとしたパケットをドロップ

【改  訂  日】2007/11/12
【変  更  点】上記09/17の修正を削除，DropQueueクラスを新規設計
【説      明】09/17の修正はDropQueueクラスで行う

----------------------------------------------------------------------

【パッケージ】jist.swans.route
【ファイル名】RouteInterface.java
【改  訂  日】2007/11/13
【変  更  点】メソッドの追加
【説      明】RouteInterface.RouteAodvインターフェース内に
　　　　　　　RREQUnicastTimeout(Object, NetMessage, MacAddress)追加

----------------------------------------------------------------------

【パッケージ】jist.swans.route
【ファイル名】RouteAodv.java
【改  訂  日】2007/11/13
【変  更  点】定数の変更と追加
　　　　　　　RouteRequestクラスにunicast(MacAddress)メソッドを追加
　　　　　　　自身のエントリのアップデート方法変更
　　　　　　　RouteTableクラスにlookup(MacAddress)メソッドの追加
　　　　　　　RouteTableEntryクラスにフィールド変数追加
　　　　　　　RouteTableEntryクラスに新規フィールド変数用処理の追加
　　　　　　　MessageQueueクラスにフィールド変数追加
　　　　　　　MessageQueueクラスに新規フィールド変数処理の追加
　　　　　　　MessageQueueクラスdequeueAndSend(NetAddress, MacAddress)の処理内容変更
　　　　　　　PrecursorSetクラスsendRERR(LinkedList, byte)の処理内容変更
　　　　　　　OutgoingSetクラスにremove(MacAddress)の追加
　　　　　　　OutgoingInfoクラスにフィールド変数追加
　　　　　　　OutgoingInfoクラスコンストラクタ処理変更，getLastMsgTime()追加
　　　　　　　RREQUnicastTimeout(Object, NetMessage, MacAddress)追加
　　　　　　　timeout()の処理内容変更
　　　　　　　broadcastHello()追加
　　　　　　　sendIpMsg(Ip, MacAddress)の処理内容変更
　　　　　　　peek(NetMessage, MacAddress)の処理内容変更
　　　　　　　send(NetMessage)の処理内容変更
　　　　　　　receive(Message, NetAddress, MacAddress, byte, NetAddress, byte, byte)処理内容変更
　　　　　　　receiveRouteRequestMessage(RouteRequestMessage, NetAddress, MacAddress, NetAddress, byte, byte)処理内容変更
　　　　　　　receiveRouteReplyMessage(RouteReplyMessage, NetAddress, MacAddress, NetAddress, byte, byte)処理内容変更
　　　　　　　receiveHelloMessage(HelloMessage, MacAddress)の追加
　　　　　　　generateRouteReplyMessage(RouteRequestMessage, boolean, RouteTableEntry)処理内容変更
　　　　　　　computeRREQTimeout(byte)処理内容変更
　　　　　　　getHops(NetAddress)の削除
【説      明】定数の値の変更と一部削除および追加
　　　　　　　RREQをユニキャスト送信するメソッドの追加
　　　　　　　自身の経路情報の更新手法の変更
　　　　　　　MACアドレスからIPアドレスを求めるメソッドの追加

jist-swans/src/driver/ndp.java
//////////////////////////////////////////////////
// JIST (Java In Simulation Time) Project
// Timestamp: <ndp.java Tue 2004/04/06 11:58:00 barr pompom.cs.cornell.edu>
//

// Copyright (C) 2004 by Cornell University
// All rights reserved.
// Refer to LICENSE for terms and conditions of use.

package driver;

import jist.swans.field.Field;
import jist.swans.field.Spatial;
import jist.swans.field.PathLoss;
import jist.swans.field.Mobility;
import jist.swans.field.Fading;
import jist.swans.field.Placement;
import jist.swans.net.NetIp;
import jist.swans.net.NetAddress;
import jist.swans.net.PacketLoss;
import jist.swans.mac.MacDumb;
import jist.swans.mac.MacAddress;
import jist.swans.radio.RadioNoiseIndep;
import jist.swans.radio.RadioInfo;
import jist.swans.app.AppHeartbeat;
import jist.swans.misc.Location;
import jist.swans.misc.Mapper;
import jist.swans.misc.Util;
import jist.swans.Constants;

import jist.runtime.JistAPI;

import java.io.*;
import jargs.gnu.*; // Download from: http://jargs.sourceforge.net

/**
 * Node Discovery Protocol macro-benchmark.
 *
 * @author Rimon Barr &lt;barr+jist@cs.cornell.edu&gt;
 * @version $Id: ndp.java,v 1.16 2004/04/06 16:07:42 barr Exp $
 */

public class ndp
{
  //////////////////////////////////////////////////
  // Constants
  //

  /** benchmark version. */
  public static final String VERSION = "0.1";

  /** benchmark mode. */
  private static final int MODE_INVALID = -1;
  /** benchmark mode. */
  private static final int MODE_TIME = 0;
  /** benchmark mode. */
  private static final int MODE_MEM = 1;
  /** benchmark mode. */
  private static final int MODE_DENSITY = 2;

  /** benchmark mode string. */
  private static final String MODE_TIME_STRING = "time";
  /** benchmark mode string. */
  private static final String MODE_MEM_STRING = "mem";
  /** benchmark mode string. */
  private static final String MODE_DENSITY_STRING = "density";

  /** benchmark binning types. */
  private static final int MODE_SPATIAL_LINEAR = 0;
  /** benchmark binning types. */
  private static final int MODE_SPATIAL_GRID   = 1;
  /** benchmark binning types. */
  private static final int MODE_SPATIAL_HIER   = 2;

  /** benchmark binning type strings. */
  private static final String MODE_SPATIAL_STRING_LINEAR = "linear";
  /** benchmark binning type strings. */
  private static final String MODE_SPATIAL_STRING_GRID   = "grid";
  /** benchmark binning type strings. */
  private static final String MODE_SPATIAL_STRING_HIER   = "hier";

  /** random waypoint pause time. */
  public static final int PAUSE_TIME = 30;
  /** random waypoint granularity. */
  public static final int GRANULARITY = 10;
  /** random waypoint minimum speed. */
  public static final int MIN_SPEED = 2;
  /** random waypoint maximum speed. */
  public static final int MAX_SPEED = 10;

  /** benchmark start time. */
  private static long startTime; 

  //////////////////////////////////////////////////
  // Static command-line helper methods
  //

  /**
   * Print benchmark version information.
   */
  private static void showVersion()
  {
    System.out.println("JiST Project: Node discovery protocol macro-benchmark v"+VERSION);
    System.out.println("Rimon Barr <barr+jist@cs.cornell.edu>, Cornell University.");
    System.out.println();
  }

  /**
   * Print benchmark command-line syntax.
   */
  private static void showUsage() 
  {
    System.out.println("Usage: ndp -m <mode> -n <num> -d <size> -t <time>");
    System.out.println("       ndp -v | -h");
    System.out.println();
    System.out.println("  -h, --help         display this help information");
    System.out.println("  -v, --version      display version information");
    System.out.println("  -m, --mode         benchmark mode: time, mem, density");
    System.out.println("  -n, --num          number of nodes");
    System.out.println("  -d, --dim          length of square field (meters)");
    System.out.println("  -t, --time         duration of simulation (seconds)");
    System.out.println("  -s, --spatial      [linear], grid:n, hier:n");
    System.out.println();
    System.out.println("eg: swans driver.ndp -m time -n 50 -d 4082 -t 900");
    System.out.println("    swans driver.ndp -m time -n 10000 -d 57735 -t 120 -s hier:9");
    System.out.println();
  }

  /**
   * Parsed command-line options.
   */
  private static class cmdlineOpts
  {
    // defaults
    /** print help. */
    public boolean help = false;
    /** print version. */
    public boolean version = false;
    /** benchmark mode. */
    public int mode = MODE_INVALID;
    /** number of nodes. */
    public long num = -1;
    /** length of field. */
    public long dim = 1;
    /** total simulation time. */
    public long time = 0;
    /** string of binning mode. */
    public String spatial = "linear";
    /** binning mode. */
    public int spatial_mode = MODE_SPATIAL_LINEAR;
    /** binning degree. */
    public int spatial_div = -1;
  }

  /**
   * Parse command-line options.
   *
   * @param args command-line parameters
   * @return parsed command-line options
   * @throws CmdLineParser.OptionException invalid option encountered
   */
  private static cmdlineOpts parseCommandLineOptions(String[] args)
    throws CmdLineParser.OptionException
  {
    // setup
    cmdlineOpts options = new cmdlineOpts();
    CmdLineParser parser = new CmdLineParser();
    CmdLineParser.Option opt_help = parser.addBooleanOption('h', "help");
    CmdLineParser.Option opt_version = parser.addBooleanOption('v', "version");
    CmdLineParser.Option opt_mode = parser.addStringOption('m', "mode");
    CmdLineParser.Option opt_num = parser.addStringOption('n', "num");
    CmdLineParser.Option opt_dim = parser.addStringOption('d', "dim");
    CmdLineParser.Option opt_time = parser.addStringOption('t', "time");
    CmdLineParser.Option opt_spatial = parser.addStringOption('s', "spatial");

    // parse
    parser.parse(args);
    if(parser.getOptionValue(opt_help)!=null) 
    {
      options.help = true;
    }
    if(parser.getOptionValue(opt_version)!=null) 
    {
      options.version = true;
    }
    if(parser.getOptionValue(opt_mode)!=null)
    {
      String mode = ((String)parser.getOptionValue(opt_mode)).toLowerCase();
      if(mode.equals(MODE_TIME_STRING.toLowerCase()))
      {
        options.mode = MODE_TIME;
      }
      else if(mode.equals(MODE_MEM_STRING))
      {
        options.mode = MODE_MEM;
      }
      else if(mode.equals(MODE_DENSITY_STRING))
      {
        options.mode = MODE_DENSITY;
      }
      else
      {
        throw new RuntimeException("unrecognized mode:"+mode);
      }
    }
    if(parser.getOptionValue(opt_num)!=null)
    {
      options.num = Long.parseLong((String)parser.getOptionValue(opt_num));
    }
    if(parser.getOptionValue(opt_dim)!=null)
    {
      options.dim = Long.parseLong((String)parser.getOptionValue(opt_dim));
    }
    if(parser.getOptionValue(opt_time)!=null)
    {
      options.time = Long.parseLong((String)parser.getOptionValue(opt_time));
    }
    if(parser.getOptionValue(opt_spatial)!=null)
    {
      options.spatial = ((String)parser.getOptionValue(opt_spatial)).toLowerCase();
    }
    if(options.spatial.startsWith(MODE_SPATIAL_STRING_LINEAR))
    {
      options.spatial_mode = MODE_SPATIAL_LINEAR;
    }
    else if(options.spatial.startsWith(MODE_SPATIAL_STRING_GRID))
    {
      options.spatial_mode = MODE_SPATIAL_GRID;
      options.spatial_div = Integer.parseInt(options.spatial.split(":")[1]);
    }
    else if(options.spatial.startsWith(MODE_SPATIAL_STRING_HIER))
    {
      options.spatial_mode = MODE_SPATIAL_HIER;
      options.spatial_div = Integer.parseInt(options.spatial.split(":")[1]);
    }
    else
    {
      throw new RuntimeException("invalid spatial structure: "+options.spatial);
    }
    String[] rest = parser.getRemainingArgs();
    return options;
  }


  //////////////////////////////////////////////////
  // simulation setup
  //

  /**
   * Initialize simulation node.
   *
   * @param i node number
   * @param field simulation field
   * @param placement node placement model
   * @param radioInfoShared shared radio information
   * @param protMap shared protocol map
   * @param plIn incoming packet loss model
   * @param plOut outgoing packet loss model
   */
  public static void createNode(int i,
      Field field, Placement placement,
      RadioInfo.RadioInfoShared radioInfoShared, Mapper protMap,
      PacketLoss plIn, PacketLoss plOut)
  {
    // create entities
    RadioNoiseIndep radio = new RadioNoiseIndep(i, radioInfoShared);
    //Mac802_11 mac = new Mac802_11(new MacAddress(i), radio.getRadioInfo());
    MacDumb mac = new MacDumb(new MacAddress(i), radio.getRadioInfo());
    NetIp net = new NetIp(new NetAddress(i), protMap, plIn, plOut);
    AppHeartbeat app = new AppHeartbeat(i, false);
    // hookup entities
    field.addRadio(radio.getRadioInfo(), radio.getProxy(), placement.getNextLocation());
    field.startMobility(radio.getRadioInfo().getUnique().getID());
    radio.setFieldEntity(field.getProxy());
    radio.setMacEntity(mac.getProxy());
    mac.setRadioEntity(radio.getProxy());
    byte intId = net.addInterface(mac.getProxy());
    mac.setNetEntity(net.getProxy(), intId);
    net.setProtocolHandler(Constants.NET_PROTOCOL_HEARTBEAT, app.getNetProxy());
    app.setNetEntity(net.getProxy());
    app.getAppProxy().run();
  }

  /**
   * Initialize simulation field.
   *
   * @param bounds size of field
   * @param spatial binning radio container
   * @param nodes number of nodes
   * @return simulation field
   */
  public static Field createSim(Location.Location2D bounds, Spatial spatial, long nodes)
  {
    // create field
    Placement placement = new Placement.Random(bounds);
    Mobility mobility = new Mobility.RandomWaypoint(bounds, PAUSE_TIME, GRANULARITY, MAX_SPEED, MIN_SPEED);
    Fading fading = new Fading.None();
    PathLoss pathloss = new PathLoss.FreeSpace();
    Field field = new Field(spatial, fading, pathloss, mobility, Constants.PROPAGATION_LIMIT_DEFAULT);
    // shared radio information
    RadioInfo.RadioInfoShared radioInfoShared = RadioInfo.createShared(
        Constants.FREQUENCY_DEFAULT, Constants.BANDWIDTH_DEFAULT,
        Constants.TRANSMIT_DEFAULT, Constants.GAIN_DEFAULT,
        Util.fromDB(Constants.SENSITIVITY_DEFAULT), Util.fromDB(Constants.THRESHOLD_DEFAULT),
        Constants.TEMPERATURE_DEFAULT, Constants.TEMPERATURE_FACTOR_DEFAULT, Constants.AMBIENT_NOISE_DEFAULT);
    // protocol mapper
    Mapper protMap = new Mapper(Constants.NET_PROTOCOL_MAX);
    protMap.mapToNext(Constants.NET_PROTOCOL_HEARTBEAT);
    // packet loss
    PacketLoss pl = new PacketLoss.Zero();
    // create nodes
    for(int i=0; i<nodes; i++)
    {
      final int PACE = 10000;
      createNode(i, field, placement, radioInfoShared, protMap, pl, pl);
      if((i+1)%PACE==0) System.out.print(" "+(i/PACE+1));
    }
    return field;
  }

  /**
   * Benchmark entry point: heartbeat benchmark.
   *
   * @param args command-line parameters
   */
  public static void main(String[] args)
  {
    try
    {
      // command line
      final cmdlineOpts options = parseCommandLineOptions(args);
      // show usage
      showVersion();
      if(options.help)
      {
        showUsage();
        return;
      }
      // show version
      if(options.version)
      {
        return;
      }
      if(options.mode==MODE_INVALID || options.num==-1)
      {
        showUsage();
        return;
      }
      // create spatial structure
      Location.Location2D bounds = new Location.Location2D(options.dim, options.dim);
      Spatial spatial = null;
      switch(options.spatial_mode)
      {
        case MODE_SPATIAL_LINEAR:
          spatial = new Spatial.LinearList(bounds);
          break;
        case MODE_SPATIAL_GRID:
          spatial = new Spatial.Grid(bounds, options.spatial_div);
          break;
        case MODE_SPATIAL_HIER:
          spatial = new Spatial.HierGrid(bounds, options.spatial_div);
          break;
        default:
          throw new RuntimeException("invalid binning type");
      }
      // run benchmark
      System.out.println("nodes   = "+options.num);
      System.out.println("size    = "+options.dim+" x "+options.dim);
      System.out.println("time    = "+options.time+" seconds");
      System.out.println("spatial = "+options.spatial);
      startTime = System.currentTimeMillis();
      System.out.print("Creating simulation nodes...");
      Field f = createSim(bounds, spatial, options.num);
      System.out.println(" done.");
      switch(options.mode)
      {
        case MODE_MEM:
          reportMem();
          JistAPI.end();
          break;
        case MODE_TIME:
          final long initTime = System.currentTimeMillis();
          JistAPI.runAt(new Runnable()
              {
                public void run()
                {
                  long endTime = System.currentTimeMillis();
                  System.out.println("seconds:"+
                    " init="+ ((initTime-startTime)/1000.0)+
                    " run="+  ((endTime-initTime)/1000.0)+
                    " total="+((endTime-startTime)/1000.0));
                }
              }, options.time*Constants.SECOND);
          JistAPI.endAt(options.time*Constants.SECOND);
          break;
        case MODE_DENSITY:
          System.out.println("Average density  = "+f.computeDensity()*1000*1000+"/km^2");
          System.out.println("Average sensing  = "+f.computeAvgConnectivity(true));
          System.out.println("Average receive  = "+f.computeAvgConnectivity(false));
          switch(options.spatial_mode)
          {
            case MODE_SPATIAL_LINEAR:
              System.out.println("Average node/bin = "+options.num);
              break;
            case MODE_SPATIAL_GRID:
              System.out.println("Average node/bin = "+options.num/(double)(options.spatial_div*options.spatial_div));
              break;
            case MODE_SPATIAL_HIER:
              System.out.println("Average node/bin = "+options.num/Math.pow(4, options.spatial_div));
              break;
            default:
              throw new RuntimeException("invalid binning type");
          }
          JistAPI.end();
          break;
        default:
          throw new RuntimeException("unknown benchmark mode");
      }
    }
    catch(CmdLineParser.OptionException e)
    {
      System.out.println("Error parsing command line: "+e.getMessage());
    }
  }

  /**
   * Print memory information.
   */
  public static void reportMem()
  {
    // report internal memory use
    System.gc();
    System.out.println("freemem:  "+Runtime.getRuntime().freeMemory());
    System.out.println("maxmem:   "+Runtime.getRuntime().maxMemory());
    System.out.println("totalmem: "+Runtime.getRuntime().totalMemory());
    System.out.println("used:     "+
        (Runtime.getRuntime().totalMemory()-Runtime.getRuntime().freeMemory()));

    // report system memory numbers
    try
    {
      byte[] b = new byte[5000];
      FileInputStream fin = new FileInputStream("/proc/self/status");
      int readbytes = fin.read(b);
      System.out.write(b, 0, readbytes);
    }
    catch(IOException ex) 
    { 
    }
  }

} // class: ndp

